### PR TITLE
classes, fields, constructors, and methods

### PR DESCRIPTION
**Pull request by mflatt**
Evolving design document is at https://github.com/mflatt/rhombus-prototype/blob/class/design/class/class.md

This is still very much a work in progress, and methods aren't in place, yet. But this overall approach to the syntax seems to be working, and field declaration and constructor-customization part is maybe all there.

### CODE REVIEW

File: design/class/class.md [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+
+Open Issues
+-----------
+
+An uncooperative custom constructor for a non-final class might return
+an instance of itself or some other subclass when called on behalf of
+the constructor for a different subclass.
+
+Class clause macros should be able to receive information about the
+fields of the enclosing class. Moving `extends` out of the clas sbody
```
Comment (Lazerbeak12345) [inline design/class/class.md:342]:
```suggestion
fields of the enclosing class. Moving `extends` out of the class body
```

Comment (mflatt) [reply design/class/class.md]: Fixed - thanks!

File: design/class/class.md [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+simple pattern-matching macro transformations, which is useful with
+`binding` and `anotation`. (When `constructor` is specified, then
+typically `binding` and `annotation` should also be specified.) The
+`entry_point` for `binding` and `annotation` is a meta-time
+expression.
+
+A class is `final` unless `nonfinal` or `extends` is present (i.e., by
+default, classes do not have subclasses). A `nonfinal` clause is
+implied when `extends` is present and `final` is not present (i.e., by
+default, a subclass can have surther subclasses).
```
Comment (sorawee) [inline design/class/class.md:130]: further

Comment (mflatt) [reply design/class/class.md]: Fixed - thanks!

**Comment by mflatt** [comment]
Now with methods and interfaces — and worth a closer look from anyone who is interested.

Probably most of the details will seem obvious, although specific syntactic choices are always debatable. The least obvious parts of the design (still with some precedents) are (1) the handling of constructor, binding, and annotation customization with subclassing, and (2) the notion of implementing an interface privately.

Remaining work includes defining interfaces for equality and printing, sorting out method access via a class name (as opposed to an object annotated with the class), and writing overview documentation and examples in the reference.

**Comment by mflatt** [comment]
Merged as c917717e710ddb626643626b85301cbf11daab0d

**Pull request closed without merging**